lib/pacemaker/pcmk_rule.c

```diff
@@ -0,0 +1,290 @@
+               const char *rule_id, crm_time_t *date)
+{
+    xmlNodePtr cib_constraints = NULL;
+    xmlNodePtr match = NULL;
+    xmlXPathObjectPtr xpath_obj = NULL;
+    char *xpath = NULL;
+    int rc = pcmk_rc_ok;
+    int max = 0;
+
+    CRM_ASSERT((out != NULL) && (data_set != NULL) && (rule_id != NULL)
+               && (date != NULL));
+
+    /* Rules are under the constraints node in the XML, so first find that. */
+    cib_constraints = pcmk_find_cib_element(data_set->input,
+                                            XML_CIB_TAG_CONSTRAINTS);
+
+    /* Get all rules matching the given ID that are also simple enough for us
+     * to check. For the moment, these rules must only have a single
+     * date_expression child and:
+     * - Do not have a date_spec operation, or
+     * - Have a date_spec operation that contains years= but does not contain
+     *   moon=.
+     *
+     * We do this in steps to provide better error messages. First, check that
+     * there's any rule with the given ID.
+     */
+    xpath = crm_strdup_printf("//rule[@id='%s']", rule_id);
+    xpath_obj = xpath_search(cib_constraints, xpath);
+    max = numXpathResults(xpath_obj);
+
+    if (max == 0) {
+        rc = ENXIO;
+        out->err(out, "No rule found with ID=%s", rule_id);
+        goto done;
+    } else if (max > 1) {
+        rc = ENXIO;
+        out->err(out, "More than one rule with ID=%s found", rule_id);
+        goto done;
+    }
+
+    free(xpath);
+    freeXpathObject(xpath_obj);
+
+    /* Next, make sure it has exactly one date_expression. */
+    xpath = crm_strdup_printf("//rule[@id='%s']//date_expression", rule_id);
+    xpath_obj = xpath_search(cib_constraints, xpath);
+    max = numXpathResults(xpath_obj);
+
+    if (max != 1) {
+        rc = EOPNOTSUPP;
+        out->err(out, "Can't check rule %s because it does not have exactly "
+                 "one date_expression", rule_id);
+        goto done;
+    }
+
+    free(xpath);
+    freeXpathObject(xpath_obj);
+
+    /* Then, check that it's something we actually support. */
+    xpath = crm_strdup_printf("//rule[@id='%s']//date_expression["
+                              "@operation!='date_spec']", rule_id);
+    xpath_obj = xpath_search(cib_constraints, xpath);
+    max = numXpathResults(xpath_obj);
+
+    if (max == 0) {
+        free(xpath);
+        freeXpathObject(xpath_obj);
+
+        xpath = crm_strdup_printf("//rule[@id='%s']//date_expression["
+                                  "@operation='date_spec' "
+                                  "and date_spec/@years "
+                                  "and not(date_spec/@moon)]", rule_id);
+        xpath_obj = xpath_search(cib_constraints, xpath);
+        max = numXpathResults(xpath_obj);
+
+        if (max == 0) {
+            rc = ENXIO;
+            out->err(out, "Rule either must not use date_spec, or use "
+                     "date_spec with years= but not moon=");
+            goto done;
```

Comment:
We could potentially still call `out->message()` in this "unsupported" case, and add a handler to the format functions for `ENXIO`.

Maybe even do the same for the other cases where we bail out early (no rule found, more than one rule found, not exactly one date expression).

Reply:
I don't really have any strong feelings on that one way or the other.  Maybe try converting one piece and see what it looks like.

Reply:
Looks fine to me. I'll point it out in the new regression test output in a sec. I don't have strong feelings here either.